src/swapRouter.ts

```diff
@@ -175,62 +179,244 @@ export abstract class SwapRouter {
+          }
+
+          calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('exactInputSingle', [exactInputSingleParams]))
+        } else {
+          const path = trade.route.path.map((token) => token.address)
+
+          const exactInputParams = [amountIn, performAggregatedSlippageCheck ? 0 : amountOut, path, recipient]
+
+          calldatas.push(SwapRouter.INTERFACE.encodeFunctionData('swapExactTokensForTokens', exactInputParams))
+        }
+      } else {
+        let acc = []
+        let j = 0
+        while (j < route.pools.length) {
+          // seek forward until finding a pool of different type
+          let section = []
+          if (route.pools[j] instanceof Pool) {
+            while (route.pools[j] instanceof Pool) {
+              section.push(route.pools[j])
+              j++
+              if (j === route.pools.length) {
+                // we've reached the end of the route
+                break
+              }
+            }
+            acc.push(section)
+          } else {
+            while (route.pools[j] instanceof Pair) {
+              section.push(route.pools[j])
+              j++
+              if (j === route.pools.length) {
+                // we've reached the end of the route
+                break
+              }
+            }
+            acc.push(section)
+          }
+        }
+
+        const firstInputToken = route.input.wrapped
+        let outputToken = firstInputToken
+
+        const getOutputOfSection = (section: (Pool | Pair)[], firstInputToken: Token) => {
+          const { inputToken } = section.reduce(
+            ({ inputToken }, pool: Pool | Pair): { inputToken: Token } => {
+              const outputToken: Token = pool.token0.equals(inputToken) ? pool.token1 : pool.token0
+              return {
+                inputToken: outputToken,
+              }
+            },
+            { inputToken: firstInputToken }
+          )
+          return inputToken
+        }
+
+        for (let i = 0; i < acc.length; i++) {
+          const isLastSectionInRoute = (i: number) => {
+            return i === acc.length - 1
+          }
+
+          const section = acc[i]
+
+          /// Previous output is now input, save this before reassigning
+          let nextInput = outputToken
+          /// Now, we get output of this section
+          outputToken = getOutputOfSection(section, outputToken)
+
+          const newRouteOriginal = new MixedRouteSDK(
+            [...section],
+            section[0].token0.equals(nextInput) ? section[0].token0 : section[0].token1,
+            outputToken
+          )
+          const newRoute = new MixedRoute(newRouteOriginal)
+          const path: string = encodeMixedRouteToPath(newRoute)
+
+          if (mixedRouteIsAllV3(newRoute)) {
+            const exactInputParams = {
+              path,
+              /// routerMustCustody is prioritized
+              recipient: isLastSectionInRoute(i) ? recipient : ADDRESS_THIS,
```

Comment:
Initially I thought that we didn't need `routerMustCustody`, but after writing more tests I found that we weren't correctly supporting the ETH output case with unwrapping the WETH. So now we bundle the routerMustCustody check into the `recipient` variable and if it's the last section in the route, we will still set it to `ADDRESS_THIS` if routerMustCustody is true. Otherwise it'll work as expected and recipient will be whatever is provided in swapOptions or sender.